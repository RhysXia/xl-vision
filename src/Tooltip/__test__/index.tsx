import { mount, ReactWrapper } from 'enzyme'
import * as React from 'react'
import Tooltip from '..'

describe('Tooltip', () => {
  let wrapper: ReactWrapper

  beforeAll(() => {
    jest.useRealTimers()
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('render', () => {
    wrapper = mount(
      <Tooltip className='popup' visible={true} content='this is a tooltip'>
        <button className='btn'>button</button>
      </Tooltip>
    )

    expect(wrapper.render()).toMatchSnapshot()
  })

  it('render:maxWidth', () => {
    wrapper = mount(
      <Tooltip className='popup' visible={true} maxWidth={100} content='this is a tooltip'>
        <button className='btn'>button</button>
      </Tooltip>
    )

    expect(wrapper.render()).toMatchSnapshot()
  })
})
