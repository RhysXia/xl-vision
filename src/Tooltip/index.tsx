import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import Popper, { PopperProps } from '../Popper'
import ConfigContext from '../ConfigProvider/ConfigContext'

export interface TooltipProps
  extends Omit<PopperProps, 'arrow' | 'popup' | 'popupClassName' | 'popupStyle'> {
  clsPrefix?: string
  content: React.ReactNode
  maxWidth?: number
  className?: string
  style?: React.CSSProperties
}

const Tooltip: React.FunctionComponent<TooltipProps> = (props) => {
  const { clsPrefix: rootClsPrefix } = React.useContext(ConfigContext)
  const {
    clsPrefix = `${rootClsPrefix}-tooltip`,
    children,
    content,
    offset = 10,
    transitionClasses = `${clsPrefix}--zoom`,
    maxWidth,
    className,
    style,
    ...others
  } = props

  const contentClasses = classnames(`${clsPrefix}__content`, {
    [`${clsPrefix}__content--with-width`]: maxWidth !== undefined
  })

  const arrow = <div className={`${clsPrefix}__arrow`} />
  const popup = (
    <div className={contentClasses} style={{ maxWidth }}>
      {content}
    </div>
  )

  const classes = classnames(clsPrefix, className)

  return (
    <Popper
      {...others}
      offset={offset}
      popupClassName={classes}
      popupStyle={style}
      arrow={arrow}
      popup={popup}
      transitionClasses={transitionClasses}
    >
      {children}
    </Popper>
  )
}

Tooltip.displayName = 'Tooltip'

Tooltip.propTypes = {
  clsPrefix: PropTypes.string,
  style: PropTypes.object,
  className: PropTypes.string,
  maxWidth: PropTypes.number,
  children: PropTypes.element.isRequired,
  offset: PropTypes.any,
  transitionClasses: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  content: PropTypes.node.isRequired
}

export default Tooltip
