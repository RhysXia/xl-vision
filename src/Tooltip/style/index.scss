@use '../../Popper/style';
@use './var' as *;
@use '../../style/mixins/arrow' as *;

$tooltip: #{$cls-prefix}-tooltip;

.#{$tooltip} {
  position: relative;

  &__content {
    box-sizing: border-box;
    padding: $tooltip-padding;
    color: $tooltip-color;
    font-size: $tooltip-font-size;
    line-height: $tooltip-line-height;
    background-color: $tooltip-bg;
    border-radius: $tooltip-border-radius;

    &--with-width {
      white-space: pre-wrap;
      text-align: justify;
      word-wrap: break-word;
      word-break: break-all;
    }
  }

  @include arrow(#{$tooltip}__arrow, $tooltip-arrow-size, $tooltip-bg);

  &--zoom {
    &-enter-active,
    &-leave-active {
      transition: $tooltip-aniamtion;
    }

    &-enter,
    &-leave-to {
      transform: scale(0);
      opacity: 0;
    }

    &-enter-to,
    &-leave {
      transform: scale(1);
      opacity: 1;
    }
  }
}
