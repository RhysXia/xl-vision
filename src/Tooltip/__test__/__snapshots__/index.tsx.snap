// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip render 1`] = `
Array [
  <div
    data-popper-escaped=""
    data-popper-placement="top"
    data-popper-reference-hidden=""
    style="z-index: 1001; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, -10px);"
  >
    <div
      class="xl-tooltip popup"
      data-placement="top"
      style="transform-origin: 0px 100%;"
    >
      <div
        class="xl-tooltip__arrow"
        data-popper-arrow=""
        style="position: absolute; left: 0px; transform: translate(0px, 0px);"
      />
      <div
        class="xl-tooltip__content"
      >
        this is a tooltip
      </div>
    </div>
  </div>,
  <button
    class="btn"
  >
    button
  </button>,
]
`;

exports[`Tooltip render:maxWidth 1`] = `
Array [
  <div
    data-popper-escaped=""
    data-popper-placement="top"
    data-popper-reference-hidden=""
    style="z-index: 1002; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, -10px);"
  >
    <div
      class="xl-tooltip popup"
      data-placement="top"
      style="transform-origin: 0px 100%;"
    >
      <div
        class="xl-tooltip__arrow"
        data-popper-arrow=""
        style="position: absolute; left: 0px; transform: translate(0px, 0px);"
      />
      <div
        class="xl-tooltip__content xl-tooltip__content--with-width"
        style="max-width: 100px;"
      >
        this is a tooltip
      </div>
    </div>
  </div>,
  <button
    class="btn"
  >
    button
  </button>,
]
`;
